File: javascriptv3/example_code/bedrock-agent-runtime/actions/invoke-flow.js
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { fileURLToPath } from "node:url";

import {
  BedrockAgentRuntimeClient,
  InvokeFlowCommand,
} from "@aws-sdk/client-bedrock-agent-runtime";

/**
 * Invokes an alias of a flow to run the inputs that you specify and return
 * the output of each node as a stream.
 *
 * @param {{
 *  flowIdentifier: string,
 *  flowAliasIdentifier: string,
 *  prompt?: string,
 *  region?: string
 * }} options
 * @returns {Promise<import("@aws-sdk/client-bedrock-agent").FlowNodeOutput>} An object containing information about the output from flow invocation.
 */
export const invokeBedrockFlow = async ({
  flowIdentifier,
  flowAliasIdentifier,
  prompt = "Hi, how are you?",
  region = "us-east-1",
}) => {
  const client = new BedrockAgentRuntimeClient({ region });

  const command = new InvokeFlowCommand({
    flowIdentifier,
    flowAliasIdentifier,
    inputs: [
      {
        content: {
          document: prompt,
        },
        nodeName: "FlowInputNode",
        nodeOutputName: "document",
      },
    ],
  });

  let flowResponse = {};
  const response = await client.send(command);

  for await (const chunkEvent of response.responseStream) {
    const { flowOutputEvent, flowCompletionEvent } = chunkEvent;

    if (flowOutputEvent) {
      flowResponse = { ...flowResponse, ...flowOutputEvent };
      console.log("Flow output event:", flowOutputEvent);
    } else if (flowCompletionEvent) {
      flowResponse = { ...flowResponse, ...flowCompletionEvent };
      console.log("Flow completion event:", flowCompletionEvent);
    }
  }

  return flowResponse;
};

// Call function if run directly
import { parseArgs } from "node:util";
import {
  isMain,
  validateArgs,
} from "@aws-doc-sdk-examples/lib/utils/util-node.js";

const loadArgs = () => {
  const options = {
    flowIdentifier: {
      type: "string",
      required: true,
    },
    flowAliasIdentifier: {
      type: "string",
      required: true,
    },
    prompt: {
      type: "string",
    },
    region: {
      type: "string",
    },
  };
  const results = parseArgs({ options });
  const { errors } = validateArgs({ options }, results);
  return { errors, results };
};

if (isMain(import.meta.url)) {
  const { errors, results } = loadArgs();
  if (!errors) {
    invokeBedrockFlow(results.values);
  } else {
    console.error(errors.join("\n"));
  }
}

File: javav2/example_code/bedrock-agents-runtime/src/main/java/com/example/bedrockagents/runtime/InvokeFlow.java
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
package com.example.bedrockagents.runtime;

// snippet-start:[bedrock-agent.java2.InvokeFlow]
import java.util.concurrent.CompletableFuture;
import software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider;
import software.amazon.awssdk.core.document.Document;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.bedrockagentruntime.BedrockAgentRuntimeAsyncClient;
import software.amazon.awssdk.services.bedrockagentruntime.model.InvokeFlowRequest;
import software.amazon.awssdk.services.bedrockagentruntime.model.InvokeFlowResponse;
import software.amazon.awssdk.services.bedrockagentruntime.model.InvokeFlowResponseHandler;
import software.amazon.awssdk.services.bedrockagentruntime.model.FlowInput;
import software.amazon.awssdk.services.bedrockagentruntime.model.FlowInputContent;
import software.amazon.awssdk.core.SdkBytes;
import software.amazon.awssdk.core.async.SdkPublisher;


/**
 * Before running this Java V2 code example, set up your development
 * environment, including your credentials.
 *
 * For more information, see the following documentation topic:
 *
 * https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/get-started.html
 */


public class InvokeFlow {

    public static String invokeFlowString(String[] args) {
        final String usage = """

        Usage:
           <flowId> <flowAliasId> <inputText>

        Where:
           flowId - An instance id value that you can obtain from the AWS Management Console.\s
           flowAliasId - An instance id value that you can obtain from the AWS Management Console.\s
           inputText - A monitoring status (true|false)""";


        if (args.length != 3) {
            System.out.println(usage);
            System.exit(1);
        }

        String flowId = args[0];
        String flowAliasId = args[1];
        String inputText = args[2];

        //Create Agent Runtime Async Client
        BedrockAgentRuntimeAsyncClient AgentClient = BedrockAgentRuntimeAsyncClient.builder()
                .credentialsProvider(DefaultCredentialsProvider.create())
                .region(Region.US_EAST_1)
                .build();
        

        //Create input prompt
        Document doc = Document.fromString(inputText);
        FlowInputContent flowInputContent = FlowInputContent.builder().document(doc).build();
        FlowInput flowInput = FlowInput.builder().nodeName("FlowInputNode").nodeOutputName("document").content(flowInputContent).build();

        //Create Invoke Flow Request
        InvokeFlowRequest invokeFlowRequest = InvokeFlowRequest.builder().flowAliasIdentifier(flowAliasId).flowIdentifier(flowId).inputs(flowInput).build();
        
        //Build a string buffer to contain all events 
        var completeResponseTextBuffer = new StringBuilder();

        //Invoke the Invoke Flow endpoint
        CompletableFuture<Void> future = AgentClient.invokeFlow(invokeFlowRequest,
        new InvokeFlowResponseHandler() {
                @Override
                public void responseReceived(InvokeFlowResponse response) {
                    System.out.println("Flow response received: " + response +"\n");
                    completeResponseTextBuffer.append(response.toString());
                }

                @Override
                @SuppressWarnings("unchecked")
                public void onEventStream(SdkPublisher publisher) {
                publisher.subscribe(event -> {
                        if (event instanceof SdkBytes) {
                            SdkBytes bytes = (SdkBytes) event;
                            System.out.println("Message: " + bytes.asUtf8String()+"\n");
                            completeResponseTextBuffer.append(bytes.asUtf8String());
                        } 
                        else {
                            System.out.println("Received event: " + event + "\n");
                            completeResponseTextBuffer.append(event.toString());
                        }
                });
                }

                @Override
                public void exceptionOccurred(Throwable throwable) {
                    System.err.println("Error occurred: " + throwable.getMessage());
                }

                @Override
                public void complete() {
                    System.out.println("Flow invocation completed");
                }
            });

        future.join();
        // Return the complete response text.
        return completeResponseTextBuffer.toString();

    }
}
// snippet-end:[bedrock-agent.java2.InvokeFlow]